client/executor/wasmtime/src/util.rs

```diff
@@ -98,3 +98,91 @@ pub(crate) fn write_memory_from(
 	memory[range].copy_from_slice(data);
 	Ok(())
 }
+
+/// Checks whether the `madvise(MADV_DONTNEED)` works as expected.
+///
+/// In certain environments (e.g. when running under the QEMU user-mode emulator)
+/// this syscall is broken.
+#[cfg(target_os = "linux")]
+fn is_madvise_working() -> std::result::Result<bool, String> {
+	unsafe {
+		// Allocate two memory pages.
+		let pointer = rustix::mm::mmap_anonymous(
+			std::ptr::null_mut(),
+			2 * 4096,
+			rustix::mm::ProtFlags::READ | rustix::mm::ProtFlags::WRITE,
+			rustix::mm::MapFlags::PRIVATE,
+		)
+		.map_err(|error| format!("mmap failed: {}", error))?;
+
+		// Dirty them both.
+		std::ptr::write_volatile(pointer.cast::<u8>(), b'A');
+		std::ptr::write_volatile(pointer.cast::<u8>().add(4096), b'B');
+
+		// Clear the first page.
+		let result_madvise = rustix::mm::madvise(pointer, 4096, rustix::mm::Advice::LinuxDontNeed)
+			.map_err(|error| format!("madvise failed: {}", error));
+
+		// Fetch the values.
+		let value_1 = std::ptr::read_volatile(pointer.cast::<u8>());
+		let value_2 = std::ptr::read_volatile(pointer.cast::<u8>().add(4096));
+
+		let result_munmap = rustix::mm::munmap(pointer, 2 * 4096)
+			.map_err(|error| format!("munmap failed: {}", error));
+
+		result_madvise?;
+		result_munmap?;
+
+		// Verify that the first page was cleared, while the second one was not.
+		Ok(value_1 == 0 && value_2 == b'B')
+	}
+}
+
+#[cfg(test)]
+#[cfg(target_os = "linux")]
+#[test]
+fn test_is_madvise_working() {
+	assert_eq!(is_madvise_working(), Ok(true));
+}
```

Comment:
Hmm... well, maybe, but I'd probably just leave this here and see if anyone complains. Maybe it'll motivate someone to actually fix QEMU. :P

Reply:
(I would complain, because I sometimes run all tests on my virtual machine that is running on Mac os :P)

Reply:
By VM do you mean an actual VM which emulates a whole machine with a full Linux running under it? Because that should work just fine and not be affected.

Reply:
Okay, fine, I've changed this test to check that it never returns an `Err` instead of it checking that madvise always works. So now it should never fail.